### PR TITLE
chore: Mono repo setup, config and base OCA package

### PR DESCRIPTION
**Pull request by JamesKEbert**
# Summary of Changes

Updates the repo to be in a mono-repo structure as described in https://hackmd.io/OyHcGdPaQyKEoGeP1AYPqA?both#New-Directory-Structure-James-amp-Akiff-Discussion-2-21-23.

This moves the existing application and code to /packages/legacy. I expect there may be some breaking changes that I did not catch, but I was able to get the app to run successfully in legacy.

I also created some base configuration for the oca package (@aries-bifold/oca), which will help with my efforts on the npm package publishing.

These updates are inspired / adapted from [Aries Framework Javascript](https://github.com/hyperledger/aries-framework-javascript).

cc: @amanji 

# Related Issues

N/A

# Pull Request Checklist

Tick all boxes below to demonstrate that you have completed the respective task. If the item does not apply to your this PR **check it anyway** to make it apparent that there's nothing to do.

- [x] All commits contain a DCO `Signed-off-by` line (we use the [DCO GitHub app](https://github.com/apps/dco) to enforce this);
- [x] Updated LICENSE-3RD-PARTY.md for any added dependencies or vendored components;
- [x] Updated documentation as needed for changed code and new or modified features;
- [x] Added sufficient [tests](../__tests__/) so that overall code coverage is not reduced.

If you have _any_ questions to _any_ of the points above, just **submit and ask**! This checklist is here to _help_ you, not to deter you from contributing!

Pro Tip 🤓

- Read our [contribution guide](../CONTRIBUTING.md) at least once; it will save you a few review cycles!
- Your PR will likely not be reviewed until all the above boxes are checked and all automated tests have passed.

_PR template adapted from the Python attrs project._

### CODE REVIEW

**Review by amanji — COMMENTED** [review]
Did we decide that `/app` would be at the root instead of `/legacy`?

**Comment by cvarjao** [comment]
Can we have only one `yarn.lock` file as well as one yarn installation bootstrap (`.yarn/*`) instead of per package?

**Comment by amanji** [comment]
> Did we decide that `/app` would be at the root instead of `/legacy`?

We can leave this alone for now

**Comment by amanji** [comment]
Looks like the build is passing now. If we fix up the merge conflicts I'd be more than happy to put my stamp on this one.

**Comment by JamesKEbert** [comment]
Awesome--I'm looking to get the conflicts resolved prior to our call tomorrow 👍

File: package.json [hunk 1]
```diff
@@ -1,57 +1,24 @@
...
-  "author": "",
-  "license": "Apache-2.0",
   "devDependencies": {
-    "npm-run-all": "^4.1.5"
+    "prettier": "^2.8.4"
   },
-  "packageManager": "yarn@3.3.1"
+  "engines": {
+    "node": ">=16.15.0 <18.0.0",
+    "yarn": "^1.22.19"
```
Comment (cvarjao) [inline package.json]: is there a reason we are downgrading from yarn 3 to yarn 1?

Comment (JamesKEbert) [reply package.json]: I wanted to have the discussion on this last week, but unfortunately you weren't able to make it to the relevant call. The biggest issue--we've declared that we want Bifold to align with AFJ structure to increase consistency and leverage the work that's already been done, which I think is a really good idea. AFJ uses [Lerna](https://lerna.js.org/) for its mono-repo management, and Lerna is not compatible with Yarn 3. 
Secondarily, I think having to use Yarn 3 for Bifold and then use Yarn 1 for AFJ and then a mix of npm & yarn 1 for applications poses headaches for developers. 
Happy to discuss further though!

**Comment by JamesKEbert** [comment]
@amanji the checks might still be failing--it appears Bifold's ci doesn't re-run checks on PR updates. My most recent commit resolves this--but maybe should be a separate PR. Do we want the app builds to run on every update? It seems like a lot, but if we don't we could end up with failing builds/tests and not know it.

**Comment by amanji** [comment]
Historically we've run builds every time for the reasons you specified (check for failing tests, etc). Which part is currently failing?

**Comment by JamesKEbert** [comment]
By adding `types: [opened, synchronize, reopened, labeled]` to the pull_request trigger of the quality.yml & main.yml it should re-trigger checks on updates to PRs (or if it was reopened, for instance). I compared this with how AFJ is setup for reference.

**Comment by JamesKEbert** [comment]
I can open a separate PR to make that change if helpful @amanji?

**Comment by amanji** [comment]
Sure. I can take a look at it. This one looks like there's still some conflicts to resolve.

**Comment by JamesKEbert** [comment]
Conflicts resolved--tests also passing locally

**Comment by JamesKEbert** [comment]
@cvarjao does the above change resolve the yarn.lock issues you were expressing?

File: package.json [hunk 2]
```diff
@@ -1,57 +1,25 @@
...
+    "prettier": "run-s prettier",
+    "test": "cd packages/legacy/core && npm test"
   },
-  "author": "",
-  "license": "Apache-2.0",
   "devDependencies": {
-    "npm-run-all": "^4.1.5"
+    "prettier": "^2.8.4"
   },
-  "packageManager": "yarn@3.3.1"
```
Comment (cvarjao) [inline package.json:56]: can we keep `packageManager` entry for now? This is managed by corepack when enabling/configuring packaging manager.

File: package.json [hunk 3]
```diff
@@ -1,57 +1,25 @@
...
-    "packages": [
-      "core",
-      "app"
-    ]
-  },
-  "engines": {
-    "node": ">=16.15.0 <18.0.0",
-    "yarn": "^3.3.1"
-  },
-  "resolutions": {
```
Comment (cvarjao) [inline package.json]: I found the resolution block required to normalize those dependencies across all packages/workspaces. In addition, it result with having multiple versions of the same dependency which cause problems with react native.

**Review by cvarjao — CHANGES_REQUESTED** [review]
I was unable to get it building and running in Android emulator.

File: package.json (hunk 2 above)
Comment (JamesKEbert) [reply package.json:56]: For posterity: we've decided to stay at yarn 3 for this PR and make switching to yarn 1 a separate item/PR

File: package.json (hunk 3 above)
Comment (JamesKEbert) [reply package.json]: Added 👍

**Comment by JamesKEbert** [comment]
@cvarjao the above commits should fix the linting--had to add the legacy package to make it work. Just awaiting the GH runner to become available again.

**Comment by TelegramSam** [comment]
@cvarjao can you verify and clear requested changes if fixed?

**Comment by cvarjao** [comment]
LGTM! We got is tested with BC Wallet, and a have a drafted PR waiting for this one:
- bcgov/bc-wallet-mobile#1012

**Review by cvarjao — APPROVED** [review]

**Review by bryce-mcmath — APPROVED** [review]

**Comment by bryce-mcmath** [comment]
@JamesKEbert One thing I didn't test (and is now unfortunately broken for me) is the git hooks flow - pre commit lint and all that. did you have to do anything special to get it working for you?

**Comment by JamesKEbert** [comment]
I missed that one--I've merged in your fix in #686, thanks for fixing! :)

**Pull request merged**